### ISSUE TITLE
CMake expected ASIOSDK location is undocumented

### ISSUE THREAD

**Issue by EvanBalster** [PortAudio/portaudio#518]
When using CMake to generate portaudio with ASIO support, the ASIO SDK directory must be a sibling of the portaudio repository.  This is not documented in the ASIO readme:
https://github.com/PortAudio/portaudio/blob/master/src/hostapi/asio/ASIO-README.txt.

I had to investigate this CMake file to suss it out:
https://github.com/portaudio/portaudio/blob/master/cmake_support/FindASIOSDK.cmake

**To Reproduce**

# Use CMake to configure PortAudio for building in Visual Studio (or any Windows compiler).
# Select `PA_USE_ASIO` in CMake's configuration options.
# Attempt to generate the project.

**Expected behavior**

CMake build instructions should be documented at least briefly in https://github.com/PortAudio/portaudio/blob/master/src/hostapi/asio/ASIO-README.txt.

Recommend CMake should support one or more of the following:

- A configuration variable for ASIOSDK_ROOT_DIR
- Also scanning for the ASIO SDK in `src/hostapi/asio`

**Actual behavior**

CMake reports missing values of ASIOSDK_ROOT_DIR and ASIOSDK_INCLUDE_DIR.  Defining these as CMake environment variables has no effect.  The CMake configuration does not provide a a way to provide these values.

**Desktop:**
 - OS: Windows
 - OS Version: 10
 - PortAudio version: main branch
 - ASIO host API

**Comment by Be-ing** [comment]
#461 searches the portaudio repository root and the directory above it for the SDK ZIP file. Alternatively the path can be explicitly specified with the CMake option `ASIO_SDK_ZIP_PATH`.

**Comment by EvanBalster** [comment]
Historical methods for building Portaudio with ASIO support involved placing the ASIO SDK in `src/hostapi/asio`.  Although it's a strange location, it is still indicated by the documentation and so I would recommend adding it to the search directories.

**Comment by RossBencina** [comment]
I agree with your "Expected behavior" @EvanBalster, the supported location(s) should be consistent between the docs, the legacy MSVC build file and CMake.  On the other hand, I agree with @Be-ing about the logical locations for the SDK.

It's not clear from @Be-ing's comment whether the SDK is expected to be zipped or unzipped. But certainly an unzipped option should be supported since it is common to patch bugs in the SDK.

**Comment by Be-ing** [comment]
The issue is that the documentation has not been updated yet. I don't think it's worth working on the docs until the new build system is made for the C++ bindings, but that can't be done until someone makes the new repo (which I have asked for several times).

**Comment by RossBencina** [comment]
Related #662

**Comment by RossBencina** [comment]
This particular issue is unrelated to C++ bindings and can be resolved independently of C++ bindings by updating the CMake script to automatically check both preferred and legacy locations for the ASIO SDK.

**Comment by EvanBalster** [comment]
If I recall correctly, the issue I previously encountered was CMake looking out-of-tree for path `../ASIOSDK` relative to the portaudio root.  Presumably on the assumption that portaudio is checked out to a folder full of libraries...

So to answer @RossBencina...  That list looks sufficient, so long as the priority order makes sense.  I would say:

1. User-supplied CMake Configuration Variable
2. Environment Variable
3. Autodetect (`src/hostapi/asio/ASIOSDK` or `../ASIOSDK`)

**Comment by RossBencina** [comment]
The path forward seems clear. We're keen to accept a patch from whoever wants to fix this.

**Comment by RossBencina** [comment]
> This issue is not relevant anymore as ASIO dependency is downloaded automatically

I disagree. For a few reasons:

1. The CMake code does still allow the use of a local zip file, it just does not search the old well-known location (hence it is a regression that needs to be fixed: the old location should be searched!).

2. Manually managing the exact SDK version (and hence the ASIO SDK download) is essential for producing reproducible builds. In most cases you do not want to just blindly pull the latest SDK version from the Steinberg site. Even our CI code downloads a specific version and caches it.

3. The ASIO SDK has known bugs, and it is typical to manage a local patched version.

**Comment by dechamps** [comment]
> Manually managing the exact SDK version (and hence the ASIO SDK download) is essential for producing reproducible builds. In most cases you do not want to just blindly pull the latest SDK version from the Steinberg site

I agree, the ASIO SDK URL should be a specific version and the CMakeLists.txt should include a hash check to guarantee the zipfile is the one that's expected. This is what I did for one of my projects: https://github.com/dechamps/ASIOUtil/blob/781482bdd272abfaed92f0002d6dc9320cc1a1db/CMakeLists.txt#L9

**Comment by dmitrykos** [comment]
> 1. The CMake code does still allow the use of a local zip file, it just does not search the old well-known location (hence it is a regression that needs to be fixed: the old location should be searched!).

@RossBencina could you please specify what `old well-known location` actually is? According to `ASIO-README.txt` it can be any, if am not mistaken. Currently PA CMake will look for existing `asiosdk` folder in the parent folder of PA source folder:

For example if I place asiosdk in the parent folder of PA:
`Found ASIO SDK: E:/SVN/audio/portaudio_git/../asiosdk`

So SDK is actually found successfully for me taking into account that predefined location of ASIO SDK folder.

> In most cases you do not want to just blindly pull the latest SDK version from the Steinberg site. 

Is there any reason for this assumption? Unless some future SDK release breaks PA ASIO completely and renders it unusable I do not see any reason to stick to some old SDK version which may have bugs which were likely fixed in the new. Moreover it seems Steinberg is not super active with this SDK with last release dated 2019th (latest pulled `asiosdk_2.3.3_2019-06-14`).

If developer needs to use some specific version of SDK then it could be hosted in the folder or can be provided as ZIP via `ASIO_SDK_ZIP_PATH`:

`cmake -DASIO_SDK_ZIP_PATH="E:/SVN/audio/portaudio_git/my_specific_asiosdk.zip"`

**Comment by dechamps** [comment]
> Is there any reason for this assumption? Unless some future SDK release breaks PA ASIO completely and renders it unusable I do not see any reason to stick to some old SDK version which may have bugs which were likely fixed in the new.

Because of the basic principle of reproducible software builds: you want builds to be hermetic and a given source tree should always compile to the same thing, regardless of external factors such as some third party updating some URL somewhere. This makes things easier to reason about (this PortAudio commit = this version of the ASIO SDK, period), ensures an update to the ASIO SDK cannot suddenly break PortAudio, and improves security against software supply chain attacks (if the download is hash-checked for integrity, which it should be).

That doesn't mean the ASIO SDK cannot be updated, it just means it will only be updated on PortAudio's own terms with an explicit PortAudio commit. Of course PortAudio users can always do the update with a local patch if they really want.

**Comment by dmitrykos** [comment]
@dechamps PA is not dealing with rocket science therefore freezing some SDK version does not make sense as progress and development move forward (see how Apple enforces this progress). I have already mentioned that if developer wants to freeze SDK at some version then there is possibility to do it by providing ZIP via `ASIO_SDK_ZIP_PATH` or by keeping offline folder.

**Comment by RossBencina** [comment]
> PA is not dealing with rocket science therefore freezing some SDK version does not make sense

It is a standard engineering practice to manage and gate update of 3rd party dependencies behind a review/test/acceptance process. Another manifestation of this is "version pinning". This is not rocket science. Once I enter a release cycle, I do not arbitrarily pull the latest versions of any dependencies, I keep using the version of dependencies that has been used throughout the UAT cycle.

> as progress and development move forward (see how Apple enforces this progress). 

I would not use Apple as a positive example. Apple enforces that "progress" as part of their forced obsolescence strategy. I have had my product magically acquire new bugs by Apple's updates -- exactly the reason *not* to pull 3rd party dependency updates without review and testing.

**Comment by RossBencina** [comment]
Adding to 19.8 milestone as it is an identified regression that the CMakeLists.txt does not check the v19.7 location (i.e. it will break existing v19.7 builds).

**Comment by RossBencina** [comment]
# Summary of PortAudio v19.7.0 ASIO SDK paths

Referencing files at: https://github.com/PortAudio/portaudio/blob/v19.7.0

## Legacy MSVC projects

- `portaudio/build/msvc/portaudio.dsp` (MSVC 6):
  hardcoded relative path `..\..\src\hostapi\asio\ASIOSDK` i.e. `portaudio\src\hostapi\asio\ASIOSDK`
    
- `portaudio/build/msvc/portaudio.vcproj` (Visual Studio 2005):
  hardcoded relative path `..\..\src\hostapi\asio\ASIOSDK` i.e. `portaudio\src\hostapi\asio\ASIOSDK`

## configure.in

- default: `/src/local/asiosdk2`
- configurable with `--with-asiodir`
  
```  
AC_ARG_WITH(asiodir,
    AS_HELP_STRING([--with-asiodir], [ASIO directory @<:@/usr/local/asiosdk2@:>@]),
    with_asiodir=$withval, with_asiodir="/usr/local/asiosdk2")
```

## doxygen docs

- `portaudio\doc\src\tutorial\compile_cmake.dox`: says to follow instructions in `compile_windows_asio_msvc`.
  *"If you want ASIO support you need to obtain the ASIO2 SDK from Steinberg and place it according to \ref compile_windows_asio_msvc. Both ASIO and the DirectX SDK are automatically searched for by the CMake script - if they are found, they will be enabled by default."*

- `portaudio\doc\src\tutorial\compile_windows_asio_msvc.dox`: *"This tutorial describes how to build PortAudio with ASIO support using MSVC *from scratch*, without an existing Visual Studio project."* Assumes the following file system layout:
    ```
        /ASIOSDK2 
        /portaudio
        /DirContainingYourVisualStudioProject
    ```

- `portaudio\doc\src\tutorial\compile_windows.dox`: Describes how to build portaudio using MSVC, by setting up the msvc project file from scratch, however the ASIO section refers to the "shipped projects" (i.e. portaudio.dsp, portaudio.vcproj) and refers to the ASIO SDK files as located within the project at Source Files > hostapi > ASIO > ASIOSDK (note that this is a project tree location not a file system path).

## CMake

PortAudio 19.7 CMake ASIO SDK path search is implemented in `portaudio\cmake_support\FindASIOSDK.cmake` which is invoked via a `find_package(ASIOSDK)` call in `CMakeLists.txt`. The path search works as follows:

### TL;DR

User can specify the ASIO SDK base directory by passing `-DASIOSDK_ROOT_DIR=C:\\path\\to\\asiosdk` when configuring CMake. If the user did not specify a path, `FindASIOSDK.cmake` searches for a directory containing `common\asio.h` using the standard CMake `find_path()` search order augmented by a set of hints. The hints comprise all directories matched by '..\as*' (?case insensitive) relative to the `portaudio` repository root directory.

Note that recent versions of CMake provide a generalized method for the user to hint/override `find_path()` searches in `FindPACKAGENAME.cmake` files by setting the `PACKAGENAME_ROOT` variable; in our case by setting the `ASIOSDK_ROOT` variable.[1]


### Detailed description

Referenced (input) CMake variable:

- `CMAKE_CURRENT_SOURCE_DIR` the directory containing the CMakeLists.txt that invoked find_package(ASIOSDK). In our case this is the root portaudio directory.

- `ASIOSDK_ROOT_DIR` the path to the ASIO SDK base directory. The user can specify `-DASIOSDK_ROOT_DIR=C:\\path\\to\\asiosdk` to CMake on the command line.

Output variables (used by portaudio/CMakeLists.txt)

- `ASIOSDK_ROOT_DIR` is the path to the ASIOSDK, computed as described below. `CMakeLists.txt` uses this variable to construct the following include directories:
    - `${ASIOSDK_ROOT_DIR}/common`
    - `${ASIOSDK_ROOT_DIR}/host`
    - `${ASIOSDK_ROOT_DIR}/host/pc`
and the following source files:
    - `${ASIOSDK_ROOT_DIR}/common/asio.cpp`
    - `${ASIOSDK_ROOT_DIR}/host/pc/asiolist.cpp`
    - `${ASIOSDK_ROOT_DIR}/host/asiodrivers.cpp`

- `ASIOSDK_FOUND` set to true by `FIND_PACKAGE_HANDLE_STANDARD_ARGS(...)` if and only if both `ASIOSDK_ROOT_DIR` and `ASIOSDK_INCLUDE_DIR` are `if()`-true; no file existance checks are performed beyond those performed by `find_path()`.


### PortAudio CMake ASIO SDK directory search algorithm

#### v19.7:

`ASIOSDK_ROOT_DIR` is compute by `find_path()` as follows: If the CMake variable `ASIOSDK_ROOT_DIR` is already set, and its value is NOT `NOTFOUND` and does not end with `-NOTFOUND`, then no search will be performed and the existing value of `ASIOSDK_ROOT_DIR` will be accepted. Otherwise, i.e, if `ASIOSDK_ROOT_DIR` is not set, or its value is `NOTFOUND` or ends in `-NOTFOUND`, `find_path(ASIOSDK_ROOT_DIR common/asio.h ...)` will search a series of well specified locations for a directory that contains `common/asio.h`. Among those locations will be the paths specified by `ASIOSDK_PATH_HINT` as `HINTS`, see the `find_path()` documentation for a very clear statement of the search order.[1]

`ASIOSDK_PATH_HINT` is constructed to contain the paths of all sibling directories to the root `portaudio` directory with names starting with `as` (?case insensitive). This is consistent with `doc\src\tutorial\compile_windows_asio_msvc.dox` which suggests the following file system layout:
```
/ASIOSDK2 
/portaudio
```

Note that `find_path()` does not search `${ASIOSDK_PATH_HINT}` first. For example `CMAKE_PREFIX_PATH` and a number of other standard paths are searched first.

`ASIOSDK_INCLUDE_DIR` is only used by `FindASIOSDK.cmake`. It serves as a buggy validity check that `${ASIOSDK_ROOT_DIR}/common/` contains `asio.h`. However `find_path(ASIOSDK_INCLUDE_DIR...)` will succeed if `asio.h` is present in any of the standard `find_path` search locations. By specifying the desired path with `PATHS`,  `${ASIOSDK_ROOT_DIR}/common` will be checked last. Possibly the author intended to specify `NO_DEFAULT_PATH`.[1]

`find_path(ASIOSDK_ROOT_DIR common/asio.h ...)` search order is as follows (note `var:...` denotes a cmake variable, `env:...` denotes an environment variable, `plat-var:...` denotes a cmake variable defined in the Platform files for the current platform):

1. `${var:ASIOSDK_ROOT}/common/asio.h`
2. `${env:ASIOSDK_ROOT}/common/asio.h`
3. `<prefix>/include/common/asio.h` and `<prefix>/include/<arch>` for each `<prefix>` in `CMAKE_PREFIX_PATH` cmake var
4. `${var:CMAKE_INCLUDE_PATH}/common/asio.h`
5. `${var:CMAKE_FRAMEWORK_PATH}/common/asio.h`
6. `<prefix>/include/common/asio.h` and `<prefix>/include/<arch>` for each `<prefix>` in `CMAKE_PREFIX_PATH` environment variable
7. `${env:CMAKE_INCLUDE_PATH}/common/asio.h`
8. `${env:CMAKE_FRAMEWORK_PATH}/common/asio.h`
9. each glob match for `portaudio/../as*/common/asio.h` (`HINTS`) (sorted in lexicographical order since cmake 3.6 --probably the opposite of what you want, probably want latest asio sdk if multiple are present)
10. `<path>/common/asio.h` for each `<path>` in `INCLUDE` and `PATH` environment variables
11. `<prefix>/include/common/asio.h` and `<prefix>/include/<arch>` for each `<prefix>` in `CMAKE_SYSTEM_PREFIX_PATH` cmake var defined in the Platform files for the current system.
12. `plat-var:CMAKE_SYSTEM_INCLUDE_PATH/common/asio.h`
13. `plat-var:CMAKE_SYSTEM_FRAMEWORK_PATH/common/asio.h`

#### current master:

First, collect candidates into the `HEADER_FILE`  variable using `file(GLOB ...)`: matching any and all of the following case insensitive patterns. (Since cmake 3.6 the resulting list of matching paths is sorted lexicographically)

- `"${var:CMAKE_CURRENT_BINARY_DIR}/asiosdk*/common/asio.h"`, `CMAKE_CURRENT_BINARY_DIR` is the root of the portaudio build tree.

- `"${var:CMAKE_PREFIX_PATH}/asiosdk*/common/asio.h"`, a quoted string, so the value of cmake variable `CMAKE_PREFIX_PATH` is substituted literally. **This is a bug** which expands to one of the following:

  - `"/asiosdk*/common/asio.h"` if `CMAKE_PREFIX_PATH` is not set.
  
  - `"C:/some/path/asiosdk*/common/asio.h"` if the user set a single prefix path using a command line option like `-DCMAKE_PREFIX_PATH=C:/some/path` (the desired result)

  - `" C:/some/path;D:/other/path/asiosdk*/common/asio.h"` an invalid path Windows path containing a semicolon. This happens if the user passed multiple entries like so: `-DCMAKE_PREFIX_PATH=C:/some/path;D:/other/path`. **[BUG!]**

- `portaudio/../asiosdk*/common/asio.h` a restricted form of the repository-relative legacy pattern `portaudio/../as*/common/asio.h`

Then the following check is performed: `if(NOT EXISTS "${HEADER_FILE}")`. `EXISTS` always interprets `HEADER_FILE` as a *single file path*. So if the `file(GLOB ...)` found multiple matches, this check fails and the `.zip` file is downloaded and used. **[BUG!]**

Note that the current master sets `ASIO_ROOT` but never checks it, so the modern CMake form `-DASIO_ROOT=C:\some\path` is not currently available as an override. The only override that is implemented is `-DCMAKE_PREFIX_PATH=C:\some;path` but if it is set to more than one path then it does not work.


[1] https://cmake.org/cmake/help/latest/command/find_path.html

**Comment by RossBencina** [comment]
[UPDATED:]Having now reviewed the v19.7.0 build systems and documentation (see my previous message, and extra detail below) here are my thoughts:

- I agree that the documentation around this could be improved. I can update the legacy MSVC project tutorials. We can update the cmake tutorial once this ticket is finalised.

- Neither v19.7.0 nor current master `CMakeLists.txt` refer to ASIO SDK at `src/hostapi/asio/ASIOSDK`. That path is only referenced by the legacy MSVC files (found in `portaudio/msvc/ in master). I don't see any reason for portaudio's cmake build system to reference that path.

- v19.7's `ASIOSDK_INCLUDE_DIR` did not provide any user-visible utility, we do not need to support that going forward.

- Both the v19.7.0 and current master `CMakeLists.txt` search for the ASIO SDK in the parent of the portaudio root directory, however they do so using different effective glob patterns (v19.7: `../as*/common/asio.h` vs. master: `../asiosdk*/common/asio.h`) and different logic: v19.7.0 used `find_path()` over a glob set and standard paths ~~and I think we should go back to a solution based on  `find_path()`. That will give much better backward compatibility than the current situation, and will allow users to use CMake's standard `PackageName_ROOT` override mechanism.~~[1]  whereas current master uses `file(GLOB ...)`. Note since `find_path()` does not support `*` wildcard matching, it is not directly usable to find wildcard directory names such as `asiosdk*`; any solution will be a combination of `file(GLOB, ...)` and `find_path()`.

- We should support the v19.7 override `ASIOSDK_ROOT_DIR` for backward compatibility.

- We should support `<PROJECTNAME>_ROOT` cmake variable and environment variable overrides as implemented by `find_path()`

- Fix bug in master: must not fail when `CMAKE_PREFIX_PATH` has multiple entries

- Fix bug in master: must not fail silently when the `asio*/common/asio.h` glob matches multiple sdk versions (printing an error and requiring the user to specify one using `-DASIOSDK_ROOT=...` is one reasonable option.

- Fix bug in master: must not fail if multiple zip files are found

- I am not sure why the package was renamed from `ASIOSDK` to `ASIO`. I'm not sure it is necessary to have changed it and I'm inclined to change it back.

- I notice that the current master `FindASIO.cmake` will write`patched_asiolists.cpp` in the ASIO SDK directory (line 68). I think we discussed placing the patched file in the build directory.

- I am not clear on why current master checks `CMAKE_CURRENT_BINARY_DIR`, I guess this is to pick up the previously downloaded .zip file

- being able to override `ASIO_SDK_ZIP_PATH` is good. should we use `ASIOSDK_...` naming convention consistently though? is `ZIP_PATH` the correct suffix?

- I am not at all comfortable with automatically downloading the ASIO SDK without explicit user opt-in. e.g `-DASIOSDK_DOWNLOAD_SDK=1`.

- It would be good to be able to specify an alternate download location: e.g. `-DASIOSDK_DOWNLOAD_URL=...`

- It would be preferable to move the download logic into a helper module.


For reference:

## PortAudio CMake code to locate ASIO SDK:

### v19.7.0

v19.7.0 source: https://github.com/PortAudio/portaudio/blob/v19.7.0/cmake_support/FindASIOSDK.cmake

`ASIOSDK_ROOT_DIR` is the ASIO SDK dir variable, computed as:

```
file(GLOB results "${CMAKE_CURRENT_SOURCE_DIR}/../as*") # ${portaudio_root}/../as* case insensitive
foreach(f ${results})
  if(IS_DIRECTORY ${f})
    set(ASIOSDK_PATH_HINT ${ASIOSDK_PATH_HINT} ${f})
  endif()
endforeach()

find_path(ASIOSDK_ROOT_DIR
  common/asio.h
  HINTS
    ${ASIOSDK_PATH_HINT}
)
```

### Current master

current master source: https://github.com/PortAudio/portaudio/blob/master/cmake/modules/FindASIO.cmake

`ASIO_ROOT` is the ASIO SDK dir variable, computed as:

```
file(GLOB HEADER_FILE
  "${CMAKE_CURRENT_BINARY_DIR}/asiosdk*/common/asio.h" # case insensitive
  "${CMAKE_PREFIX_PATH}/asiosdk*/common/asio.h" # case insensitive
  # The old build systems before PortAudio 19.8 used to look for the ASIO SDK
  # in the same parent directory as the source code repository. This is no
  # longer advised or documented but kept for backwards compatibility.
  "${CMAKE_CURRENT_SOURCE_DIR}/../asiosdk*/common/asio.h" # ${portaudio_root}/../asiosdk* case insensitive
)
if(NOT EXISTS "${HEADER_FILE}")
  ... (download zip fuke) ...
endif()

get_filename_component(HEADER_DIR "${HEADER_FILE}" DIRECTORY)
get_filename_component(ASIO_ROOT "${HEADER_DIR}" DIRECTORY)
```

NOTES:

1. On Windows and Mac, `file(GLOB ...)` is always case insensitive irrespective of the file system's case sensitivity.[2]

2. Since CMake 3.6, `file(GLOB ...)` returns results sorted lexicographically, this will affect the detected asio sdk if multiple SDKs are found.[2]

## CMake find_path() function search order

See the cmake `find_path()` documentation for details and nuances. [1]

In the following `<PackageName>` is the case-preserved package name (i.e. `ASIOSDK` in PA v19.7, `ASIO` in current master), and `<PACKAGENAME>` is the upper-cased package name (v19.7: `ASIOSDK`, master: `ASIO`).

1.
  a. `<PackageName>_ROOT` CMake variable (since CMake 3.12)
  b. `<PACKAGENAME>_ROOT` CMake variable (since 3.27)
  c. `<PackageName>_ROOT` environment variable (since CMake 3.12)
  d. `<PACKAGENAME>_ROOT` environment variable (since 3.27)
  
2. Search paths specified in cmake-specific cache variables (unless explicitly disabled by flag or var):
 - `<prefix>/include` and if `CMAKE_LIBRARY_ARCHITECTURE` is set `<prefix>/include/<arch>`, for each prefix in `CMAKE_PREFIX_PATH`
 - `CMAKE_INCLUDE_PATH`
 - `CMAKE_FRAMEWORK_PATH`
 
 Note: these variables are intended to be specified on command line using `-DVAR=value`
 
3. Search paths specified in cmake-specific environment variables `CMAKE_PREFIX_PATH`, `CMAKE_INCLUDE_PATH`, `CMAKE_FRAMEWORK_PATH` (same as (2.) but env vars)

4. Search paths specified by `HINTS` option

5. Search standard system env vars (unless explicitly disabled by flag or var)
 - the directories in `INCLUDE` and `PATH` env vars
 
6. Search cmake variables defined in Platform files for the current system (unless explicitly disabled by flag or var):
 - `<prefix>/include` and if `CMAKE_LIBRARY_ARCHITECTURE` is set `<prefix>/include/<arch>`, for each prefix in `CMAKE_SYSTEM_PREFIX_PATH`
 - `CMAKE_SYSTEM_INCLUDE_PATH`
 - `CMAKE_SYSTEM_FRAMEWORK_PATH`

7. Search paths specified by the `PATHS` option.

Note that there are also non-local mechanisms to exclude/ignore certain paths.

[1] https://cmake.org/cmake/help/latest/command/find_path.html

[2] https://cmake.org/cmake/help/latest/command/file.html#glob

**Comment by RossBencina** [comment]
Phil and I discussed this today and reached the following conclusions:

***Automated download of the ASIO SDK will be removed***, for the following reasons:

1. Legal: The commercial ASIO SDK licensing terms are not compatible with any arrangement that could be construed as PortAudio developers distributing the ASIO SDK as an integrated part of PortAudio (see clause 2.3, p. 2 [1]); or as acting as intermediaries between the PortAudio user and Steinberg. For example, clause 2.7, p. 2 [1] can be read as requiring PortAudio developers warrant that PortAudio user's are in compliance with all ASIO SDK license terms. The upshot being that relationships and mechanisms need to be structured such that the user is directly downloading and licensing the ASIO SDK from Steinberg. That means that the PortAudio user needs to take responsibility for sourcing the ASIO SDK and understanding their legal relationship with Steinberg.

2. Legal: The ASIO SDK is now dual-licensed with a GPL option. Using GPL code in an application has non-trivial licensing implications that are materially different from PortAudio's MIT license terms. We do not want to be the cause of any confusion in this regard.

3. Technical: Automatic download of software from a third-party website is inherently unreliable (network and servers can fail) and brittle (URLs can change). Providing a simple download feature as is currently implemented can only be considered a quick and dirty hack. It is not a robust build capability.

4. Security Risk: Downloading executables or source code from a third-party website without appropriate precautions is always dangerous. To date, no effort has been made to check file hashes or otherwise harden the download script.

5. Network Overhead: The current implementation will re-download the ASIO SDK each time the /build directory is created (perhaps this is why our Windows CI is slow?). This is another indicator that the current implementation is a quick hack.

6. Automatic download without informed user consent is not viable. Opt-in download with informed consent requires users to be made aware of, and consent to, all of the previous points.

In principle, 3,4,5 and 6 could be addressed, but at that point PortAudio would be implementing what amounts to a bespoke package manager for automatically pulling the ASIO SDK. Even if someone were to implement such a system, for the legal reasons given above I do not believe that such an effort should be housed within the PortAudio project. And in any case, we are not going to delay the v19.8 release while this gets resolved. So for v19.8 we will remove the auto-download code.

---

To close out this issue for PortAudio 19.8, I think three PRs are needed:

- [ ] Fix ASIO SDK search to be backward compatible with v19.7 (e.g. ?use find_path, ?glob ../as*, legacy ASIOSDK_ROOT_DIR override) and also support standard CMAKE path override <PACKAGENAME>_ROOT behavior. Ideally, prefer standard CMake mechanisms to glob. Review use of current master CMAKE_PREFIX_PATH usage, which is definitely buggy and looks to me like a non-standard use of CMAKE_PREFIX_PATH ("prefix" usually applies to FHS paths, it means that <prefix>/include will be searched.) This PR should include the necessary documentation changes that indicate where PortAudio searches by default, and how to tell CMake where the SDK is.

- [ ] Fix the code in portaudio/cmake/modules/FindAsio.cmake that patches the Windows ASIO SDK bug to (1) check tha lpdrv is allocated using array operator new prior to applying the patch, (2) always put the patched file in the build dir, (3) ideally, set up the patched file as a generated file that depends on the unpatched file. (The goal is to make this fully compatible with users who have supplied an already-patched copy of the ASIO SDK.) PR: https://github.com/PortAudio/portaudio/pull/1173

- [ ] Remove ASIO SDK download functionality, including zipfile unzipping functionality and the ASIO_SDK_ZIP_PATH variable. If building PortAudio with ASIO is enabled and CMake can't find an ASIO SDK, report an error and point to the documentation. This PR could/should fix CI to download the SDK.

I suggest that we develop and merge 1, and 2 first so that when the download stops there is a working and recommended way for PortAudio users to proceed.

In addition, I think we should review whether the package should be named "ASIO" (current master) or changed back to "ASIOSDK" (v19.7).


[1] https://download.steinberg.net/sdk_downloads/Steinberg%20ASIO%20Licensing%20Agreement.pdf

**Comment by RossBencina** [comment]
@dechamps wrote:
> ASIO SDK URL should be a specific version and the CMakeLists.txt should include a hash check to guarantee the zipfile is the one that's
> expected. This is what I did for one of my projects:
> https://github.com/dechamps/ASIOUtil/blob/781482bdd272abfaed92f0002d6dc9320cc1a1db/CMakeLists.txt#L9

Thanks for the useful reference. FetchContent certainly seems to be the modern CMake approach:

https://cmake.org/cmake/help/latest/module/FetchContent.html
https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html#fetchcontent-and-find-package-integration

When we remove the download capability from PortAudio, CMake portaudio consumers who want to automatically download from Steinberg can do something like the following:

```
# NOTE: running the following code will download the Steinberg ASIO SDK
# which is publicly downloadable but, at the time of writing, is subject to either:
#  (1) A non-OSI-approved commercial license agreement
#      see: https://www.steinberg.net/developers/prorietary-sdk/
# or
#  (2) The GNU General Public License, version GPLv3
#      see: https://www.steinberg.net/developers/asiosdk-open/

# quoting: https://github.com/dechamps/ASIOUtil/blob/781482bdd272abfaed92f0002d6dc9320cc1a1db/CMakeLists.txt#L9
include(FetchContent)
set(FETCHCONTENT_QUIET FALSE)
FetchContent_Declare(asiosdk
	URL "https://download.steinberg.net/sdk_downloads/asiosdk_2.3.3_2019-06-14.zip"
	URL_HASH SHA1=33c44f0c919a7218413ea81e06a66bb8b47be292
	USES_TERMINAL_DOWNLOAD TRUE  # https://gitlab.kitware.com/cmake/cmake/issues/18238#note_440475
)
FetchContent_GetProperties(asiosdk)
if(NOT asiosdk_POPULATED)
  FetchContent_Populate(asiosdk)
endif()
```

Then:

```
# asiosdk_SOURCE_DIR is set by FetchContent_GetProperties() or FetchContent_Populate(asiosdk)
# ASIOSDK_ROOT_DIR (or maybe ASIO_ROOT or ASIOSDK_ROOT tbc) is read by PortAudio's FindASIO.cmake
set(ASIOSDK_ROOT_DIR asiosdk_SOURCE_DIR)
find_package(portaudio)
```

**Comment by RossBencina** [comment]
PR 1 of 3: https://github.com/PortAudio/portaudio/pull/1173

Please review.

**Comment by RossBencina** [comment]
Is anyone with CMake experience able to review #1173 please?
